File: src/test/java/pages/purchasePages/PurchaseInvoiceTransactionsPage.java
package pages.purchasePages;

import com.microsoft.playwright.FrameLocator;
import com.microsoft.playwright.Locator;
import com.microsoft.playwright.options.LoadState;
import com.microsoft.playwright.options.WaitForSelectorState;
import org.junit.Assert;
import pages.commonPages.BasePage;

public class PurchaseInvoiceTransactionsPage extends BasePage {

    Locator pageTitle = page.locator("#PageTitle");
    Locator filterBtn = page.locator("#FilterButtonId");
    FrameLocator updateDeclarationFrame = getFrameByDialogTitle("Beyanname Güncelleme");
    FrameLocator createCountingFrame = getFrameByDialogTitle("Sayım Oluşturma");
    FrameLocator updateCountingFrame = getFrameByDialogTitle("Sayım Güncelleme");
    FrameLocator stockProcess = getFrameByDialogTitle("Stoğa Al");
    FrameLocator orderProcess = getFrameByDialogTitle("Sipariş İşlemleri");
    Locator invoiceNo = page.locator("#FilterInvoiceNo");
    Locator checkBox = page.locator("input[type='checkbox'][name^='InvoiceGridId']");
    Locator checkboxDeclaration = page.locator("#checkboxDeclaration");
    FrameLocator copyFrame = page.frameLocator("iframe[src*='ProductFilterBeforeCopy']");

    /**
     * Fatura İşlemleri sayfasını doğrular
     */
    public void verifyPurchaseInvoiceTransactionPage() {
        verifyTextElementUseTrim("Fatura İşlemleri",pageTitle);
    }
    /**
     * Fatura İşlemleri Fatura No alanı doldurur.
     */
    public void fillInvoiceNo() {
        clickElement(invoiceNo);
        String invoiceNumber = getString("Fatura No");
        invoiceNo.fill(invoiceNumber);
        //invoiceNo.fill("ER6788888"); //örnek data
    }
    /**
     * Fatura İşlemleri Fatura No ile sorgular.
     */
    public void searchForInvoiceNo(){
        clickElement(filterBtn);
    }
    /**
     * Fatura İşlemleri Fatura No doğrular.
     */
    public void verifyInvoiceNo(){
        String invoiceContent = page.locator("td[data-field-name='InvoiceNo']").textContent();
        String invoiceNumber = getString("Fatura No");
        Assert.assertEquals(invoiceContent,invoiceNumber);
    }
    /**
     * Fatura İşlemleri sorgu ile gelen Fatura no checkbox işaretler.
     */
    public void clickCheckboxForDeclaration() {
        clickElement(checkBox);
    }
    /**
     * Fatura İşlemleri Beyanname oluşturur. Beyanname Güncelleme frame açılır.
     */
    public void createDeclaration(){
        checkboxDeclaration.nth(0).click();
        popUpConfirmationProcess();
    }
    /**
     * Beyanname Güncelleme frame sayım tabına geçer.
     */
    public void selectCountingTab() {
        page.waitForLoadState(LoadState.DOMCONTENTLOADED);
        Locator proformaTab = updateDeclarationFrame.locator(".k-item.k-state-default").nth(2);
        proformaTab.scrollIntoViewIfNeeded();
        proformaTab.waitFor(new Locator.WaitForOptions().setState(WaitForSelectorState.ATTACHED));
        proformaTab.click(new Locator.ClickOptions().setForce(true));
    }
    /**
     * Sayım Oluşturma frame açar.
     */
    public void creatingCount(){
        updateDeclarationFrame.locator("a.k-grid-DeclarationCountGridIdAddNew").click();
    }
    /**
     * Sayım Oluşturma frame açıklama alanını random doldurur.
     */
    public void fillDescriptionField(){
        int randomNumber = generateRandomNumber();
        String formatted = String.format("SYM-%05d", randomNumber);
        createCountingFrame.locator("#Description").fill(formatted);
    }
    /**
     * Sayım Oluşturma frame kaydeder.
     */
    public void saveCountDescription(){
        createCountingFrame.locator("#SaveBtn").click();
    }
    /**
     * Sayım Güncelleme frame açar.
     */
    public void editCountingProcess() {

        updateDeclarationFrame.locator("#Edit").click();
    }
    /**
     * Sayım Güncelleme frame sayıma gönder işlemi yapar.
     */
    public void sendForCount(){
        updateCountingFrame.locator("#sendCountingClickButton").click();
        updateCountingFrame.locator(".ajs-button.ajs-ok").click();
    }
    /**
     * Sayım Güncelleme frame talebi onaylayana kopyala işlemini yapar.
     */
    public void copyRequestToApprover(){
        updateCountingFrame.locator("#btnCopyRequestedToApproved").click();
    }
    /**
     * Talebi onaylayana kopyala işlemini tamamlar.
     */
    public void  copyingProcess(){
        copyFrame.locator("#Copy").click();
        copyFrame.locator(".ajs-button.ajs-ok").click();
    }
    /**
     * Sayım Güncelleme frame sayım tamamlar.
     */
    public void completingCount(){
        updateCountingFrame.locator("#completeButton").click();
        updateCountingFrame.locator(".ajs-button.ajs-ok").click();
    }
    /**
     * Sayım Güncelleme frame kaydeder.
     */
    public void saveUpdateCount(){
        updateCountingFrame.locator("#SaveBtn").click();
    }
    /**
     * Beyanname Güncelleme frame Sevkiyat Dışı Tut checkbox iişaretler.
     */
    public void checkExcludeShipping() {

        updateDeclarationFrame.locator("#yes_ExcludeDispatch").click();
    }
    /**
     * Beyanname Güncelleme frame kaydeder.
     */
    public void saveDeclarationUpdate(){
        page.waitForLoadState(LoadState.DOMCONTENTLOADED);
        Locator saveBtn = updateDeclarationFrame.locator("#SaveBtn");
        saveBtn.scrollIntoViewIfNeeded();
        saveBtn.click();
    }
    /**
     * Fatura İşlemleri sayfası Beyanname Sistem No ile Beyanname Güncelleme frame açar.
     */
    public void openUpdateDeclarationFrame() {

        page.waitForSelector("a[title='Beyanname Sistem No']");
        page.locator("a[title='Beyanname Sistem No']").click();
    }
    /**
     * Beyanname Güncelleme frame stoğa al frame açar.
     */
    public void openPutInStockFrame(){
        updateDeclarationFrame.locator("#StockEntry").click();
    }
    /**
     * Stoğa Al frame Gümrük Beyanname No belli standarda göre doldurur.
     */
    public void fillCustomsDeclarationNoField(){
        String customsNo = generateCustomHouseNo();
        stockProcess.locator("#CustomHouseNo").fill(customsNo);
    }
    /**
     * Stoğa Al frame Gümrük Tarihi seçer.
     */
    public void selectCustomsDate(){
        stockProcess.locator(".k-icon.k-i-calendar").nth(0).click();
        stockProcess.locator(".k-link.k-nav-today").nth(0).click();
    }
    /**
     * Stoğa Al frame Yurda Giriş No doldurur.
     */
    public void fillDormitoryEntryNo(){
        int randomNo = generateRandomNumber();
        String dormitoryEntryNo = String.format("DEN-%05d", randomNo);
        stockProcess.locator("#DeclarationEntryNo").fill(dormitoryEntryNo);
    }
    /**
     * Stoğa Al frame Yurtiçi Giriş Tarihi seçer.
     */
    public void selectDomesticEntryDate(){
        stockProcess.locator(".k-icon.k-i-calendar").nth(1).click();
        stockProcess.locator(".k-link.k-nav-today").nth(1).click();
    }
    /**
     * Stoğa Al frame Rejim No seçer.
     */
    public void selectRegimeNo(){
        stockProcess.locator("span.k-dropdown-wrap").nth(0).click();
        stockProcess.locator("#RegimeNoSourceId_listbox li").nth(1).click();
    }
    /**
     * Stoğa Al frame bilgileri kaydeder.
     */
    public void savePutInStockProcess(){
        stockProcess.locator("#btnCountAndSave").click();
        Locator stockSuccessMessage = page.locator("ajs-message.ajs-success.ajs-visible");
        if (stockSuccessMessage.isVisible()){
            stockSuccessMessage.click();
        }
    }
    /**
     * Fatura İşlemleri sipariş no ile Sipariş İşlemleri frame açar.
     */
    public void openOrderProcessFrame() {

        page.waitForSelector("a[title='Sipariş']");
        page.locator("a[title='Sipariş']").click();
    }
    /**
     * Sipariş İşlemleri frame sipariş tamamla işlemini yapar.
     */
    public void completeOrder(){
        orderProcess.locator("#CompleteOrderBtn").click();
        Locator message = orderProcess.locator(".ajs-button.ajs-ok");
        if (message.isVisible()){
            message.click();
        }
    }
    /**
     * Sipariş İşlemleri frame success mesajlarını ve frame kapatır.
     */
    public void closeOrderProcessFrame() {
        Locator successPopup = page.locator(".ajs-message.ajs-success").last();

        if (successPopup.isVisible()) {
            successPopup.click();
        } else {
            System.out.println("mesajlar kapatıldı");
        }
        page.locator("div.k-window-actions a.k-window-action").nth(0).click();
    }
}

File: RetailTRUI.Tests/Pages/Purchasing/PurchaseInvoiceTransactionsPage.cs
using RetailTRUI.Tests.Pages.Common;

namespace RetailTRUI.Tests.Pages.Purchasing;

/// <summary>
/// Purchase Invoice Transactions page object for managing invoice processes
/// Includes declaration, counting, and stock entry operations
/// </summary>
public class PurchaseInvoiceTransactionsPage : BasePage
{
    private ILocator PageTitle => Page.Locator("#PageTitle");
    private ILocator FilterBtn => Page.Locator("#FilterButtonId");
    private IFrameLocator UpdateDeclarationFrame => GetFrameByDialogTitle("Beyanname Güncelleme");
    private IFrameLocator CreateCountingFrame => GetFrameByDialogTitle("Sayım Oluşturma");
    private IFrameLocator UpdateCountingFrame => GetFrameByDialogTitle("Sayım Güncelleme");
    private IFrameLocator StockProcess => GetFrameByDialogTitle("Stoğa Al");
    private IFrameLocator OrderProcess => GetFrameByDialogTitle("Sipariş İşlemleri");
    private ILocator InvoiceNo => Page.Locator("#FilterInvoiceNo");
    private ILocator CheckBox => Page.Locator("input[type='checkbox'][name^='InvoiceGridId']");
    private ILocator CheckboxDeclaration => Page.Locator("#checkboxDeclaration");
    private IFrameLocator CopyFrame => Page.FrameLocator("iframe[src*='ProductFilterBeforeCopy']");

    public async Task VerifyPurchaseInvoiceTransactionPageAsync()
    {
        var actualText = await PageTitle.TextContentAsync();
        actualText?.Trim().Should().Be("Fatura İşlemleri");
    }

    public async Task SearchByInvoiceNumberAsync()
    {
        await FillInvoiceNoAsync();
        await SearchForInvoiceNoAsync();
        await VerifyInvoiceNoAsync();
    }

    private async Task FillInvoiceNoAsync()
    {
        await InvoiceNo.ClickAsync();
        var invoiceNumber = GetString("Fatura No");
        await InvoiceNo.FillAsync(invoiceNumber);
    }

    private async Task SearchForInvoiceNoAsync()
    {
        await FilterBtn.ClickAsync();
    }

    private async Task VerifyInvoiceNoAsync()
    {
        var invoiceContent = await Page.Locator("td[data-field-name='InvoiceNo']").TextContentAsync();
        var invoiceNumber = GetString("Fatura No");
        invoiceContent.Should().Be(invoiceNumber);
    }

    public async Task OpenInvoiceUpdateFrameAsync()
    {
        await ClickCheckboxForDeclarationAsync();
        await CreateDeclarationAsync();
    }

    private async Task ClickCheckboxForDeclarationAsync()
    {
        await CheckBox.ClickAsync();
    }

    private async Task CreateDeclarationAsync()
    {
        await CheckboxDeclaration.Nth(0).ClickAsync();
        await PopUpConfirmationProcessAsync();
    }

    public async Task CompletingCountingProcessAsync()
    {
        await SelectCountingTabAsync();
        await CreatingCountAsync();
        await FillDescriptionFieldAsync();
        await SaveCountDescriptionAsync();
    }

    private async Task SelectCountingTabAsync()
    {
        await Page.WaitForLoadStateAsync(LoadState.DOMContentLoaded);
        var countingTab = UpdateDeclarationFrame.Locator(".k-item.k-state-default").Nth(2);
        await countingTab.ScrollIntoViewIfNeededAsync();
        await countingTab.WaitForAsync(new LocatorWaitForOptions { State = WaitForSelectorState.Attached });
        await countingTab.ClickAsync(new LocatorClickOptions { Force = true });
    }

    private async Task CreatingCountAsync()
    {
        await UpdateDeclarationFrame.Locator("a.k-grid-DeclarationCountGridIdAddNew").ClickAsync();
    }

    private async Task FillDescriptionFieldAsync()
    {
        var randomNumber = GenerateRandomNumber();
        var formatted = $"SYM-{randomNumber:D5}";
        await CreateCountingFrame.Locator("#Description").FillAsync(formatted);
    }

    private async Task SaveCountDescriptionAsync()
    {
        await CreateCountingFrame.Locator("#SaveBtn").ClickAsync();
    }

    public async Task EditCountingProcessAsync()
    {
        await EditCountingAsync();
        await SendForCountAsync();
        await CopyRequestToApproverAsync();
        await CopyingProcessAsync();
        await CompletingCountAsync();
        await SaveUpdateCountAsync();
    }

    private async Task EditCountingAsync()
    {
        await UpdateDeclarationFrame.Locator("#Edit").ClickAsync();
    }

    private async Task SendForCountAsync()
    {
        await UpdateCountingFrame.Locator("#sendCountingClickButton").ClickAsync();
        await UpdateCountingFrame.Locator(".ajs-button.ajs-ok").ClickAsync();
    }

    private async Task CopyRequestToApproverAsync()
    {
        await UpdateCountingFrame.Locator("#btnCopyRequestedToApproved").ClickAsync();
    }

    private async Task CopyingProcessAsync()
    {
        await CopyFrame.Locator("#Copy").ClickAsync();
        await CopyFrame.Locator(".ajs-button.ajs-ok").ClickAsync();
    }

    private async Task CompletingCountAsync()
    {
        await UpdateCountingFrame.Locator("#completeButton").ClickAsync();
        await UpdateCountingFrame.Locator(".ajs-button.ajs-ok").ClickAsync();
    }

    private async Task SaveUpdateCountAsync()
    {
        await UpdateCountingFrame.Locator("#SaveBtn").ClickAsync();
    }

    public async Task ExcludeOutOfShippingAndSaveAsync()
    {
        await CheckExcludeShippingAsync();
        await SaveDeclarationUpdateAsync();
    }

    private async Task CheckExcludeShippingAsync()
    {
        await UpdateDeclarationFrame.Locator("#yes_ExcludeDispatch").ClickAsync();
    }

    private async Task SaveDeclarationUpdateAsync()
    {
        await Page.WaitForLoadStateAsync(LoadState.DOMContentLoaded);
        var saveBtn = UpdateDeclarationFrame.Locator("#SaveBtn");
        await saveBtn.ScrollIntoViewIfNeededAsync();
        await saveBtn.ClickAsync();
    }

    public async Task PutInStockProcessAsync()
    {
        await OpenUpdateDeclarationFrameAsync();
        await OpenPutInStockFrameAsync();
        await FillCustomsDeclarationNoFieldAsync();
        await SelectCustomsDateAsync();
        await FillDormitoryEntryNoAsync();
        await SelectDomesticEntryDateAsync();
        await SelectRegimeNoAsync();
        await SavePutInStockProcessAsync();
        await SaveDeclarationUpdateAsync();
    }

    private async Task OpenUpdateDeclarationFrameAsync()
    {
        await Page.WaitForSelectorAsync("a[title='Beyanname Sistem No']");
        await Page.Locator("a[title='Beyanname Sistem No']").ClickAsync();
    }

    private async Task OpenPutInStockFrameAsync()
    {
        await UpdateDeclarationFrame.Locator("#StockEntry").ClickAsync();
    }

    private async Task FillCustomsDeclarationNoFieldAsync()
    {
        var customsNo = GenerateCustomHouseNo();
        await StockProcess.Locator("#CustomHouseNo").FillAsync(customsNo);
    }

    private async Task SelectCustomsDateAsync()
    {
        await StockProcess.Locator(".k-icon.k-i-calendar").Nth(0).ClickAsync();
        await StockProcess.Locator(".k-link.k-nav-today").Nth(0).ClickAsync();
    }

    private async Task FillDormitoryEntryNoAsync()
    {
        var randomNo = GenerateRandomNumber();
        var dormitoryEntryNo = $"DEN-{randomNo:D5}";
        await StockProcess.Locator("#DeclarationEntryNo").FillAsync(dormitoryEntryNo);
    }

    private async Task SelectDomesticEntryDateAsync()
    {
        await StockProcess.Locator(".k-icon.k-i-calendar").Nth(1).ClickAsync();
        await StockProcess.Locator(".k-link.k-nav-today").Nth(1).ClickAsync();
    }

    private async Task SelectRegimeNoAsync()
    {
        await StockProcess.Locator("span.k-dropdown-wrap").Nth(0).ClickAsync();
        await StockProcess.Locator("#RegimeNoSourceId_listbox li").Nth(1).ClickAsync();
    }

    private async Task SavePutInStockProcessAsync()
    {
        await StockProcess.Locator("#btnCountAndSave").ClickAsync();
        
        var stockSuccessMessage = Page.Locator("ajs-message.ajs-success.ajs-visible");
        if (await stockSuccessMessage.IsVisibleAsync())
        {
            await stockSuccessMessage.ClickAsync();
        }
    }

    public async Task CompleteOrderProcessAsync()
    {
        await OpenOrderProcessFrameAsync();
        await CompleteOrderAsync();
        await CloseOrderProcessFrameAsync();
    }

    private async Task OpenOrderProcessFrameAsync()
    {
        await Page.WaitForSelectorAsync("a[title='Sipariş']");
        await Page.Locator("a[title='Sipariş']").ClickAsync();
    }

    private async Task CompleteOrderAsync()
    {
        await OrderProcess.Locator("#CompleteOrderBtn").ClickAsync();
        
        var message = OrderProcess.Locator(".ajs-button.ajs-ok");
        if (await message.IsVisibleAsync())
        {
            await message.ClickAsync();
        }
    }

    private async Task CloseOrderProcessFrameAsync()
    {
        var successPopup = Page.Locator(".ajs-message.ajs-success").Last;

        if (await successPopup.IsVisibleAsync())
        {
            await successPopup.ClickAsync();
        }
        
        await Page.Locator("div.k-window-actions a.k-window-action").Nth(0).ClickAsync();
    }
}
